tests/python_package_test/test_dual.py

```diff
@@ -0,0 +1,46 @@
+"""Tests for dual GPU+CPU support."""
+
+import os
+import pytest
+
+import lightgbm as lgb
+import numpy as np
+from lightgbm.basic import LightGBMError
+
+
+@pytest.mark.skipif(
+    os.environ.get("LIGHTGBM_TEST_DUAL_CPU_GPU", None) is None,
+    reason="Only run if appropriate env variable is set",
+)
+def test_cpu_works():
+    """If compiled appropriately, the same installation will support both GPU and CPU."""
+    data = np.random.rand(500, 10)
+    label = np.random.randint(2, size=500)
+    validation_data = train_data = lgb.Dataset(data, label=label)
+
+    param = {"num_leaves": 31, "objective": "binary", "device": "cpu"}
+    # This will raise an exception if it's an unsupported device:
+    lgb.train(param, train_data, 10, valid_sets=[validation_data])
+
+
+@pytest.mark.skipif(
+    os.environ.get("LIGHTGBM_TEST_DUAL_CPU_GPU", None) is None,
+    reason="Only run if appropriate env variable is set",
+)
+def test_gpu_works():
+    """If compiled appropriately, the same installation will support both GPU and CPU."""
+    data = np.random.rand(500, 10)
+    label = np.random.randint(2, size=500)
+    validation_data = train_data = lgb.Dataset(data, label=label)
+
+    try:
+        param = {"num_leaves": 31, "objective": "binary", "device": "gpu"}
+        lgb.train(param, train_data, 10, valid_sets=[validation_data])
+    except LightGBMError as e:
+        if str(e) == "No OpenCL device found":
+            # This is fine, it means there's no OpenCL device available,
+            # and OpenCL device is only searched for if we successfully
+            # loaded OpenCL GPU backend.
+            pass
```

Comment:
Could you elaborate more on this? Why do we need this?

Reply:
This is run when the wheel is compiled for both CPU and GPU. It is explicitly requesting GPU usage.

There are two possibilities:

1. An OpenCL driver is available.
2. An OpenCL driver is not available.

In the first case, no exception is raised, test passes: we asked for GPU, got GPU, everything is good.

In the second case, a LightGBMError is raised... but because it mention "No OpenCL device found", that tells us that it tried to find a "GPU" (really, a OpenCL implementation). We asked for OpenCL, and tried to find it, so test should pass too.

Reply:
Yeah, thanks, I understand the general meaning of this case. I was wondering about our concrete CI services. Where is this workaround needed: at Appveyor or Azure Pipelines? Looks like in both cases `opencl_runtime_18.1_x64_setup.msi` will be installed and integrated GPU will be used, won't it?

Reply:
It's needed anytime that test is run and OpenCL is not installed. It's possible that is nowhere, at the moment?

Reply:
Yeah, I think we can remove it because we have the same `if` guards (`if ($env:TASK -eq "bdist")`) for installing OpenCL and running dual tests. Also, `pytest`'s `skipif` with env variable protects ordinary users from running that test.

Moreover, we do not need the second `if` and explicit command to run one new test file because that test file will be included in `$tests` variable. 
https://github.com/microsoft/LightGBM/blob/332559c0f84c2a23ec01c07737814b0aeb0ce474/.ci/test_windows.ps1#L64-L70
The only thing we need is to set `$env:LIGHTGBM_TEST_DUAL_CPU_GPU = "1"` before running tests.

Reply:
I've made this change.  test_dual.py now runs only once, with the environment variable set to "1".

Reply:
Thank you very much for all your work!